### PR TITLE
Retire pipeline/:id/apps

### PR DESCRIPTION
**Pull request by appleton**
- [x] `pipelines:info`
- [x] `pipelines:diff`
- [x] `pipelines:promote`

### CODE REVIEW

**Comment by appleton** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;jwJ1kWXE&#x2F;1448-verify-app-filtering-endpoint-meets-our-needs"> Verify app filtering endpoint meets our needs</a>

**Comment by gudmundur** [comment]
whoa, I can't believe this is happening!

File: lib/api.js [hunk 1]
```diff
@@ -48,6 +49,40 @@ function removeCoupling(heroku, app) {
...
 
+function getPipelineCouplings(heroku, pipelineId) {
+  return heroku.request({
+    method: 'GET',
+    path: `/pipelines/${pipelineId}/pipeline-couplings`,
+    headers: { 'Accept': V3_HEADER }
+  });
+}
+
+function getAppFitler(heroku, appIds) {
```
Comment (gudmundur) [inline lib/api.js]: `getAppFitler` => `getAppFilter`.

File: lib/api.js [hunk 2]
```diff
@@ -56,3 +91,11 @@ exports.deleteCoupling = deleteCoupling;
 exports.createCoupling = createCoupling;
 exports.updateCoupling = updateCoupling;
 exports.removeCoupling = removeCoupling;
+
+exports.getPipelineCouplings = getPipelineCouplings;
+
+exports.getAppFitler = getAppFitler;
```
Comment (gudmundur) [inline lib/api.js]: See above.

File: lib/api.js [hunk 3]
```diff
@@ -48,6 +49,40 @@ function removeCoupling(heroku, app) {
...
+    headers: { 'Accept': FILTERS_HEADER },
+    body: { in: { id: appIds } }
+  });
+}
+
+function getPipelineApps(heroku, pipelineId) {
+  return getPipelineCouplings(heroku, pipelineId).then((couplings) => {
+    const appIds = couplings.map((coupling) => coupling.app.id);
+
+    return getAppFitler(heroku, appIds).then((apps) => {
```
Comment (gudmundur) [inline lib/api.js]: See above.

File: lib/api.js [hunk 4]
```diff
@@ -48,6 +49,40 @@ function removeCoupling(heroku, app) {
            .then(coupling => deleteCoupling(heroku, coupling.id));
 }
 
+function getPipelineCouplings(heroku, pipelineId) {
```
Comment (gudmundur) [inline lib/api.js]: 1. Drop the `Pipeline` infix. The other wrappers here don't include it.
2. Should we call it `getCouplings` or `listCouplings`?

File: lib/api.js [hunk 5]
```diff
@@ -48,6 +49,40 @@ function removeCoupling(heroku, app) {
...
+function getAppFitler(heroku, appIds) {
+  return heroku.request({
+    method: 'POST',
+    path: `/filters/apps`,
+    headers: { 'Accept': FILTERS_HEADER },
+    body: { in: { id: appIds } }
+  });
+}
+
+function getPipelineApps(heroku, pipelineId) {
```
Comment (gudmundur) [inline lib/api.js]: Similar to above:

1. I think we should keep the `Pipeline` infix here, as `Coupling` implies pipeline above.

File: lib/api.js [hunk 6]
```diff
@@ -48,6 +49,40 @@ function removeCoupling(heroku, app) {
...
+  return getPipelineCouplings(heroku, pipelineId).then((couplings) => {
+    const appIds = couplings.map((coupling) => coupling.app.id);
+
+    return getAppFitler(heroku, appIds).then((apps) => {
+      const couplingsByAppId = couplings.reduce((memo, coupling) => {
```
Comment (gudmundur) [inline lib/api.js]: No `groupBy` anywhere in ES6? No dependency we have that does that?

File: lib/api.js [hunk 7]
```diff
@@ -48,6 +49,40 @@ function removeCoupling(heroku, app) {
...
+        memo[coupling.app.id] = coupling;
+        return memo;
+      }, {});
+
+      apps.forEach((app) => app.coupling = couplingsByAppId[app.id]);
```
Comment (gudmundur) [inline lib/api.js]: Hah, so we've moved the stitching from API to here. Is this a pattern we'd like to retain, or do we want to refactor?

File: lib/api.js (hunk 6 above)
Comment (appleton) [inline lib/api.js]: Nope!

File: lib/api.js (hunk 7 above)
Comment (appleton) [inline lib/api.js]: I don't think we have a choice really. We need to associate apps to pipelines and that happens by joining on couplings

File: lib/api.js (hunk 5 above)
Comment (appleton) [inline lib/api.js]: I actually think we keep it here - `getApps` or `listApps` feels like it will get a full app list rather than a list scoped to my pipeline. I will change it to `listPipelineApps` though

File: lib/api.js (hunk 6 above)
Comment (gudmundur) [inline lib/api.js]: :(

File: lib/api.js [hunk 8]
```diff
@@ -48,6 +49,40 @@ function removeCoupling(heroku, app) {
...
+    body: { in: { id: appIds } }
+  });
+}
+
+function listPipelineApps(heroku, pipelineId) {
+  return listCouplings(heroku, pipelineId).then((couplings) => {
+    const appIds = couplings.map((coupling) => coupling.app.id);
+
+    return getAppFilter(heroku, appIds).then((apps) => {
+      const couplingsByAppId = couplings.reduce((memo, coupling) => {
```
Comment (gudmundur) [inline lib/api.js]: Can you extract a `groupBy` helper here? Reductions can be hard to parse and it'll simplify the code.

File: lib/api.js (hunk 7 above)
Comment (gudmundur) [inline lib/api.js]: Would you mind changing this to a `map`? To me `forEach` implies not having side effects.

File: lib/api.js (hunk 8 above)
Comment (appleton) [inline lib/api.js]: Done, though this isn't actually `groupBy` because the values set are not collections. I've called it `keyBy` after https://lodash.com/docs#keyBy

**Comment by gudmundur** [comment]
:+1:
